crates/driver/src/domain/competition/solution/mod.rs

```diff
@@ -180,48 +180,57 @@ impl Solution {
-                    for input in interaction.inputs.iter() {
-                        sum_inputs += self
-                            .reference_amount(input)
-                            .ok_or(VerificationError::AssetFlow)?;
-                    }
-                    for output in interaction.outputs.iter() {
-                        sum_outputs += self
-                            .reference_amount(output)
-                            .ok_or(VerificationError::AssetFlow)?;
-                    }
-                }
-                Interaction::Liquidity(interaction) => {
-                    sum_inputs += self
-                        .reference_amount(&interaction.input)
-                        .ok_or(VerificationError::AssetFlow)?;
-                    sum_outputs += self
-                        .reference_amount(&interaction.output)
-                        .ok_or(VerificationError::AssetFlow)?;
-                }
-            }
-        }
+        // Sum settlement inputs across all interactions.
+        let inputs = self
+            .interactions
+            .iter()
+            .flat_map(|interaction| interaction.inputs().into_iter());
+        let sum_inputs = inputs
+            .fold(Some(eth::U256::zero()), |result, input| {
+                result.and_then(|sum| Some(sum + self.reference_amount(&input)?))
+            })
+            .ok_or(VerificationError::AssetFlow)?;
+
+        // Sum settlement outputs across all interactions.
+        let outputs = self
+            .interactions
+            .iter()
+            .flat_map(|interaction| interaction.outputs().into_iter());
+        let sum_outputs = outputs
+            .fold(Some(eth::U256::zero()), |result, output| {
+                result.and_then(|sum| Some(sum + self.reference_amount(&output)?))
+            })
+            .ok_or(VerificationError::AssetFlow)?;
+
+        // Check that the settlement doesn't steal tokens from the contract.
         if sum_inputs >= sum_outputs {
             Ok(())
         } else {
             Err(VerificationError::AssetFlow)
         }
     }
 
+    /// Check that internalized interactions only use trusted tokens.
     fn verify_internalization(
         &self,
-        _auction: &competition::Auction,
+        auction: &competition::Auction,
     ) -> Result<(), VerificationError> {
-        // TODO Will be done in a follow-up PR.
-        // Check that internalized interactions use trusted tokens. This requires
-        // checking the internalized interactions in the solution against the
-        // trusted tokens in the auction to make sure there's no foul play.
-        Ok(())
+        if self
+            .interactions
+            .iter()
+            .filter(|interaction| interaction.internalize())
+            .all(|interaction| {
+                interaction
+                    .inputs()
+                    .iter()
+                    .chain(interaction.outputs().iter())
```

Comment:
I think we only require that the `input` token is trusted. That is the token that would end up in the settlement contract instead of some liquidity pool so we want to be sure that it's a good token.
It's okay for the `output` token to be not trusted because that means we would get rid of a potential shit coin.